### ISSUE TITLE
One shot Datum

### ISSUE THREAD

**Issue by samhh** [nullpub/datum#8]
Would it be possible to add a variant of Datum that doesn't have any notion of refreshing? It's still superior to other types for semantic clarity, but having to cover the inapplicable refresh cases isn't very ergonomic.

**Comment by baetheus** [comment]
I believe you could write a fold-like combinator that handles that case:

```ts
import * as D from "@nll/datum/Datum";

const simplifiedFold = <A, B>(onNone: () => B, onSome: (a: A) => B) => D.fold(onNone, onNone, onSome, onSome);

const datum1: D.Datum<string> = D.initial;
const datum1: D.Datum<string> = D.initial;

**Comment by baetheus** [comment]
I'm not sure what you mean by inapplicable though. Do you mean that you aren't using it? In frontend webapp work, which is where I use this datatype most often, the refresh type is useful for showing loading spinners. The type might not be terribly useful if you don't need to know about data being refreshed though.

**Comment by samhh** [comment]
We already make use of a few helper functions like that where I work. The trouble is that it's semantically unclear if a refresh case is something that really needs to be handled because `Datum` as-is is the only type we use regardless of whether we need to handle the prospect of refreshes or not (I'd consider using a plain `Option` even more unclear, having used that previously).

At the moment, `Datum` has four members:

- `Initial`
- `Pending`
- `Refresh a`
- `Replete a`

And `DatumEither` essentially has two more, at least when you're at the point of pattern matching over it.

For us, it's very common to want to cover three cases:

- `Initial`
- `Pending`
- `Replete a` (or the split `Either` equivalent)

So, given a `DatumEither`, we have to supply six callbacks where really only four should be needed, and the developer has to go wandering off to check if there's a circumstance in which we do ever refresh the data after the initial load. It's unergonomic. Better than things were before we adopted this library, but I think it could be better still with a more diversified set of types available. (Oh, how I wish we could define sum types as cleanly as in Haskell!)

**Comment by baetheus** [comment]
I'm not sure how you're having to supply all those callbacks. There is an ergomonic function called `squash` exported from`@nll/datum/DatumEither` that looks like it does what you want it to [here](https://github.com/nullpub/datum/blob/16546b7bcc843a38809e6075c3069422ed7451f7/src/DatumEither.ts#L292). I suppose I could look into adding another adt that suits your needs, but I'll have to think about whether I'd ever use it. It's unlikely that I'd want to implement something that I won't dogfood myself. That said, if you wanted to add a new adt to this library via a PR I'd definitely welcome that and offer as much guidance and feedback as possible. My only real requirement is that is passes the build rules (100% test coverage, docs autogenerate, it doesn't add any new dependencies). If you're interested in this let me know.

**Comment by baetheus** [comment]
I decided to see how long it would take me to do a quick copy/paste/modify implementation of your one shot datum idea. Turns out only 10 minutes with tests! I don't have my yubikey on me so I'm not able to publish a version tonight but I'll likely kick out an rc release tomorrow. For now you can check out the docs [here](https://nullpub.github.io/datum/modules/OneShot.ts.html) to see if it does what you're looking for. Incidentally, I didn't implement OneShotEither and I didn't put much effort into the name. Maybe you've got some better ideas?

**Comment by baetheus** [comment]
This was resolved by 184a8ed8e49043de069e004fb19b44d70fe87c3d and released and published to [npm](https://www.npmjs.com/package/@nll/datum).

**Comment by samhh** [comment]
Fantastic! Many thanks :smiley:
